### DISCUSSION TITLE
Plugin - What are $app and $conf and how are they used?

### DISCUSSION THREAD

**Discussion by ajmetz** [mojolicious/mojo#1803] [Q&A] [answered]
Hello, 
I've seen that when making a plugin, 
in addition to $self, a subroutine mandates $app and $conf.
What are these, how are they used, and why are they not explained in the Mojolicious::Plugin base controller documentation?

**Comment by daleif** [comment]
Where exactly do you see that?

In `https://docs.mojolicious.org/Mojolicious/Plugin` we have

```
sub register ($self, $app, $conf) {
```

but it does not seem to be mandated if you compare with `https://docs.mojolicious.org/Mojolicious/Plugin/TagHelpers.txt`

But then again the latter does not use signatures.

I'm guessing the `$conf` part is optional to each Plugin, some need extra configurations passed on, others do not. The `TagHelper` example also shows what `$app` is used for.

**Comment by jhthorsen** [comment, marked as answer]
When calling `$app->plugin('My::Cool::Plugin', \%config)`, the `$app` instance (https://docs.mojolicious.org/Mojolicious) and the hash-ref `%config` will be passed on as `$config`.

Instead of `%config` you can pass on any reference or object, but not a plain scalar.

**Reply by ajmetz** [reply]
Thanks for your excellent replies.

I was busy using Data Dumper, and found:
$app to be the object corresponding to my mojolicious web app (in my case known as MojoTest and having the moniker mojo_test).
$conf is empty in my case - presumably because it has not been used (output was: $VAR1 = {}).
$self was the object corresponding to my plugin - MojoTest::Plugin::PluginName

Thanks to your post @jhthorsen - I now have an example of how I might populate $conf.
Thanks to your post @daleif - I had an example of usage of $app.

I assume it's all just Perl still, and so I could pass on anything else I need after these initial two values. Of course, presumably anything I need could be included and accessed through a hash reference or object like in this case the hashref \%config.

**Reply by jhthorsen** [reply]
No, you can’t pass in “anything”. If the third value is a plain scalar (string/number), then all the following arguments will be used to generate a hash-ref. By this, I think I would suggest to always send in a hash-ref or pair of key/values, which again would be turned into a hash-ref inside your plugin’s register() method.
